In [17]:
def confussion_matrix(objective_class, real_classes, predicted_classes):
    TP = 0
    TN = 0
    FP = 0
    FN = 0

    real_classes = list(real_classes)
    predicted_classes = list(predicted_classes)

    for i in range(len(real_classes)):
        if real_classes[i] == objective_class:
            if predicted_classes[i] == objective_class:
                TP += 1
            else:
                FN += 1
        else:
            if predicted_classes[i] == objective_class:
                FP += 1
            else:
                TN += 1
    # Print matrix
    df_matrix = pd.DataFrame({
        'Positive': [TP, FP],
        'Negative': [FN, TN]
    }, index=['Positive', 'Negative'])
    print(df_matrix)
    return TP, TN, FP, FN

def evaluation_metrics(TP, TN, FP, FN):
    accuracy = (TP + TN) / (TP + TN + FP + FN)
    precision = TP / (TP + FP)
    recall = TP / (TP + FN)
    f1_score = 2 * (precision * recall) / (precision + recall)
    return accuracy, precision, recall, f1_score


In [9]:
# Test
clase_objetivo = 'Neg'
real = pd.Series(['Pos','Pos','Neg','Neg','Neg','Pos','Pos','Pos','Pos','Pos','Neg','Pos','Neg','Pos','Neg'])
hipotesis = ['Pos','Neg','Neg','Neg','Pos','Pos','Pos','Neg','Neg','Pos','Pos','Neg','Neg','Neg','Pos']

TP, TN, FP, FN = confussion_matrix(clase_objetivo, real, hipotesis)
accuracy, precision, recall, f1_score = evaluation_metrics(TP, TN, FP, FN)
print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1 Score: {f1_score}')

          Positive  Negative
Positive         3         3
Negative         5         4 

Accuracy: 0.4666666666666667
Precision: 0.375
Recall: 0.5
F1 Score: 0.42857142857142855


In [10]:
import pandas as pd

df = pd.read_csv('diabetes.csv')
print(df.info())
df.sample(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   preg    768 non-null    int64  
 1   plas    768 non-null    int64  
 2   pres    768 non-null    int64  
 3   skin    768 non-null    int64  
 4   insu    768 non-null    int64  
 5   mass    768 non-null    float64
 6   pedi    768 non-null    float64
 7   age     768 non-null    int64  
 8   class   768 non-null    object 
dtypes: float64(2), int64(6), object(1)
memory usage: 54.1+ KB
None


preg  plas  pres  skin  insu  mass   pedi  age            class
171     8   197    74     0     0  25.9  1.191   39  tested_positive
661     8   196    76    29   280  37.5  0.605   57  tested_positive
188     3   176    86    27   156  33.3  1.154   52  tested_positive
665     9   119    80    35     0  29.0  0.263   29  tested_positive
471     1   112    80    45   132  34.8  0.217   24  tested_negative

In [11]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

X = df.drop('class', axis=1)
y = df['class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [18]:
# Model 1
model1 = DecisionTreeClassifier()
model1.fit(X_train, y_train)
y_pred = model1.predict(X_test)
TP, TN, FP, FN = confussion_matrix('tested_positive', y_test, y_pred)
accuracy, precision, recall, f1_score = evaluation_metrics(TP, TN, FP, FN)
print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1 Score: {f1_score}')

          Positive  Negative
Positive        23        27
Negative        27        77
Accuracy: 0.6493506493506493
Precision: 0.46
Recall: 0.46
F1 Score: 0.46


In [19]:
# Model 2
model2 = DecisionTreeClassifier(max_depth=3)
model2.fit(X_train, y_train)
y_pred = model2.predict(X_test)
TP, TN, FP, FN = confussion_matrix('tested_positive', y_test, y_pred)
accuracy, precision, recall, f1_score = evaluation_metrics(TP, TN, FP, FN)
print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1 Score: {f1_score}')

          Positive  Negative
Positive        33        17
Negative        28        76
Accuracy: 0.7077922077922078
Precision: 0.5409836065573771
Recall: 0.66
F1 Score: 0.5945945945945946


In [20]:
# Model 3
model3 = DecisionTreeClassifier(min_samples_leaf=5)
model3.fit(X_train, y_train)
y_pred = model3.predict(X_test)
TP, TN, FP, FN = confussion_matrix('tested_positive', y_test, y_pred)
accuracy, precision, recall, f1_score = evaluation_metrics(TP, TN, FP, FN)
print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1 Score: {f1_score}')

          Positive  Negative
Positive        25        25
Negative        24        80
Accuracy: 0.6818181818181818
Precision: 0.5102040816326531
Recall: 0.5
F1 Score: 0.5050505050505051


In [21]:
# Model 4
model4 = DecisionTreeClassifier(min_samples_split=5)
model4.fit(X_train, y_train)
y_pred = model4.predict(X_test)
TP, TN, FP, FN = confussion_matrix('tested_positive', y_test, y_pred)
accuracy, precision, recall, f1_score = evaluation_metrics(TP, TN, FP, FN)
print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1 Score: {f1_score}')

          Positive  Negative
Positive        23        27
Negative        28        76
Accuracy: 0.6428571428571429
Precision: 0.45098039215686275
Recall: 0.46
F1 Score: 0.45544554455445546


In [22]:
# Model 5
model5 = DecisionTreeClassifier(max_leaf_nodes=5)
model5.fit(X_train, y_train)
y_pred = model5.predict(X_test)
TP, TN, FP, FN = confussion_matrix('tested_positive', y_test, y_pred)
accuracy, precision, recall, f1_score = evaluation_metrics(TP, TN, FP, FN)
print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print(f'F1 Score: {f1_score}')


          Positive  Negative
Positive        16        34
Negative         7        97
Accuracy: 0.7337662337662337
Precision: 0.6956521739130435
Recall: 0.32
F1 Score: 0.4383561643835616


El modelo que obtuvo la mayor exactitud fue el Modelo 5, con un valor de 0.7338. Sin embargo, el modelo que obtuvo la mayor precisión fue el Modelo 2, con un valor de 0.5410. Por otro lado, el modelo que obtuvo el mayor recall fue el Modelo 3, con un valor de 0.6600.

Esto significa que el Modelo 5 tuvo la mayor proporción de predicciones correctas en general, mientras que el Modelo 2 tuvo la mayor proporción de predicciones correctas entre las instancias clasificadas como positivas. Por su parte, el Modelo 3 tuvo la mayor proporción de instancias positivas correctamente identificadas.

En resumen, cada modelo tuvo un rendimiento destacado en una métrica específica, lo que indica que cada uno se desempeñó mejor en diferentes aspectos de la clasificación. Por lo tanto, la elección del modelo más adecuado dependerá de la métrica que se considere más importante en el contexto del problema en cuestión.
